From 631dfc99b6bdbee82411f3d353fc70544f2cd8f1 Mon Sep 17 00:00:00 2001
From: Pier Luigi Fiorini <pierluigi.fiorini@liri.io>
Date: Wed, 24 May 2017 08:01:46 +0200
Subject: [PATCH] File existence check on Flatpak and OSTree

OSTree uses mtime 0 for all checkouts in order to maximize sharing of objects.
(i.e. two identical source files with different mtimes should result in the
same hardlinked file in the ostree checkout)

However Qbs thinks the file doesn't exist with mtime 0 so we need to rely
on a different way for OSTree.

OSTree nowadays is not only used by Flatpak, but also some Linux
distributions like Fedora Silverblue and QtOTA itself.

Change-Id: Ia2efc269dc6eb1862b4015bdcce7f9207f3a5efa

diff --git a/src/lib/corelib/tools/fileinfo.cpp b/src/lib/corelib/tools/fileinfo.cpp
index b37f20382..029e8a878 100644
--- a/src/lib/corelib/tools/fileinfo.cpp
+++ b/src/lib/corelib/tools/fileinfo.cpp
@@ -358,12 +358,15 @@ FileInfo::FileInfo(const QString &fileName)
     if (stat(fileName.toLocal8Bit(), &m_stat) == -1) {
         m_stat.st_mtime = 0;
         m_stat.st_mode = 0;
+        m_exists = false;
+    } else {
+        m_exists = true;
     }
 }
 
 bool FileInfo::exists() const
 {
-    return m_stat.st_mtime != 0;
+    return m_exists;
 }
 
 FileTime FileInfo::lastModified() const
diff --git a/src/lib/corelib/tools/fileinfo.h b/src/lib/corelib/tools/fileinfo.h
index fa101fff9..1014a1d58 100644
--- a/src/lib/corelib/tools/fileinfo.h
+++ b/src/lib/corelib/tools/fileinfo.h
@@ -97,6 +97,7 @@ private:
 #   error unknown platform
 #endif
     InternalStatType m_stat;
+    bool m_exists;
 };
 
 bool removeFileRecursion(const QFileInfo &f, QString *errorMessage);
-- 
2.17.1

